lsp/nls/src/requests/completion.rs

```diff
@@ -21,42 +104,149 @@ pub fn handle_completion(
+                        TermKind::Declaration(_, _, ValueState::Known(body_id))
+                            if name_id == i.id
+                                && find_record_contract(&linearization.linearization, body_id)
+                                    .is_some() =>
+                        {
+                            // unwrapping is safe here because of the pattern condition.
+                            let contract =
+                                find_record_contract(&linearization.linearization, body_id)
+                                    .unwrap();
+                            let fields = match &contract.types {
+                                Types(AbsType::Record(row)) => extract_ident(row),
+                                _ => Vec::with_capacity(0),
+                            };
+                            Some((fields, i.ty.clone()))
+                        }
+
+                        // Get record fields from lexical scoping
+                        TermKind::Declaration(_, _, ValueState::Known(body_id))
+                            if name_id == i.id =>
+                        {
+                            find_record_fields(&linearization.linearization, body_id)
+                                .map(|idents| Some((idents, i.ty.clone())))
+                                .unwrap_or(None)
+                        }
+
+                        _ => None,
+                    }
+                })
+                .collect()
+        }
+
+        // variable name completion
+        Some(..) | None => linearization
+            .get_in_scope(&item)
+            .iter()
+            .filter_map(|i| match i.kind {
+                TermKind::Declaration(ref ident, _, _) => Some((vec![ident.clone()], i.ty.clone())),
+                _ => None,
+            })
+            .collect::<Vec<_>>(),
+    };
+
+    // O(n^2)
+    fn remove_duplicates(items: &Vec<CompletionItem>) -> Vec<CompletionItem> {
+        let mut seen: Vec<CompletionItem> = Vec::with_capacity(0);
+        for item in items {
+            if seen
+                .iter()
+                .find(|seen_item| seen_item.label == item.label)
+                .is_none()
+            {
+                seen.push(item.clone())
+            }
+        }
+
+        seen
+    }
+
+    let in_scope: Vec<_> = in_scope
         .iter()
-        .filter_map(|i| match i.kind {
-            TermKind::Declaration(ref ident, _, _) => Some((ident.clone(), i.ty.clone())),
-            _ => None,
-        })
-        .map(|(ident, _)| CompletionItem {
-            label: ident.into_label(),
-            ..Default::default()
+        .map(|(idents, _)| {
+            idents
+                .iter()
+                .map(|ident| CompletionItem {
+                    label: ident.into_label(),
+                    ..Default::default()
+                })
+                .collect::<Vec<_>>()
         })
+        .flatten()
```

Comment:
Can we merge this `.map().flatten()` into a single `.flat_map()`?
```suggestion
        .flat_map(|(idents, _)| {
            idents
                .iter()
                .map(|ident| CompletionItem {
                    label: ident.into_label(),
                    ..Default::default()
                })
                .collect::<Vec<_>>()
        })
```